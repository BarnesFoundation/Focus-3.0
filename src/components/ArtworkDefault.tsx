import React, { Fragment, useEffect, useRef, useState } from "react";
import classnames from "classnames";
import { isAndroid } from "react-device-detect";

import { LANGUAGE_EN } from "../constants";
import withTranslation, {
  LanguageOptionType,
  WithTranslationState,
} from "./withTranslation";
import { ArtworkObject, ArtWorkRecordsResult } from "../types/payloadTypes";
import { EmailForm } from "./EmailForm";
import ProgressiveImage from "react-progressive-image";
import { LanguageDropdown } from "./LanguageDropdown";
import { Share } from "./Share";
import { ContentBlock } from "./ContentBlock";
import google_logo from "../images/google_translate.svg";
import { ScanButton } from "./ScanButton";

type ArtworkDefaultProps = {
  artwork: ArtworkObject;
  result: ArtWorkRecordsResult;
  onSelectLanguage: (selectedLanguage: LanguageOptionType) => void;
  selectedLanguage: LanguageOptionType;
  emailCaptured: boolean;
  showEmailForm: boolean;
  onSubmitEmail: (email: string, callback?: (args?: any) => void) => void;
} & WithTranslationState;

export const ArtworkDefaultComponent: React.FC<ArtworkDefaultProps> = ({
  artwork,
  result,
  onSelectLanguage,
  selectedLanguage,
  emailCaptured,
  showEmailForm,
  onSubmitEmail,
  langOptions,
  getTranslation,
}) => {
  // Component state
  const [imgLoaded, setImgLoaded] = useState(false);
  const [formOpen, setFormOpen] = useState(false);
  const startYCard2 = useRef<number>();
  const startYTrigger = useRef<number>();

  /** Updates state that email was captured and submits it to the server session */
  const handleSubmitEmail = (email: string) => {
    onSubmitEmail(email, () => setFormOpen(false));
  };

  const handleClick = () => {
    if (!formOpen) setFormOpen(true);
  };

  const handleStartSwipe = (event: TouchEvent) => {
    const touch = event.changedTouches[0];
    startYCard2.current = touch.pageY;
  };

  const handleSwipeDismiss = (event: TouchEvent) => {
    const touch = event.changedTouches[0];
    const endY = touch.pageY;
    if (endY > startYCard2.current) setFormOpen(false);
  };

  useEffect(() => {
    if (!imgLoaded || !showEmailForm) return;

    // Add handlers for card-2 swipe down interaction
    const content = document.getElementById("card-content");
    content.addEventListener("touchstart", handleStartSwipe);
    content.addEventListener("touchend", handleSwipeDismiss);
    content.addEventListener("wheel", (e) => {
      if (e.deltaY < 0) {
        setFormOpen(false);
      }
    });

    // Disable scrolling inside element
    const card2 = document.getElementById("card-2");
    card2.addEventListener("touchmove", (e: TouchEvent) => e.preventDefault());

    // Trigger to open card-2 once bottom of content is reached
    const trigger = document.getElementById("card-2__trigger");
    trigger.addEventListener("wheel", (e: WheelEvent) => {
      if (e.deltaY > 0) {
        setFormOpen(true);
      }
    });
    trigger.addEventListener("touchstart", (e: TouchEvent) => {
      startYTrigger.current = e.changedTouches[0].pageY;
    });
    trigger.addEventListener("touchend", (e: TouchEvent) => {
      if (e.changedTouches[0].pageY < startYTrigger.current) setFormOpen(true);
    });
  }, [artwork, imgLoaded, showEmailForm]);

  return (
    <div
      id="container"
      className={classnames(
        "exhibition-obj",
        isAndroid ? "sm-container" : "ios-container"
      )}
    >
      <div id="card-content" className="exhibition-obj__content">
        {!imgLoaded ? (
          <div style={{ visibility: "hidden" }}>
            <img
              className="card-img-result"
              src={artwork.url}
              alt="match_image"
              onLoad={({ target: img }) => setImgLoaded(true)}
            />
          </div>
        ) : (
          <Fragment>
            <div id="card-1" className="exhibition-obj__content__result">
              <div className="row">
                <div className="artwork-top-bg">
                  <img
                    className="card-img-top"
                    src={artwork.bg_url}
                    alt="match_image_background"
                    aria-hidden={true}
                  />
                </div>
                <div className="col-12 col-md-12">
                  <div
                    id="result-card"
                    className="card"
                    data-title={artwork.title}
                    data-artist={artwork.artist}
                    data-id={artwork.id}
                    data-invno={artwork.invno}
                    data-nodesc-invno={
                      !artwork.shortDescription ? artwork.invno : ""
                    }
                  >
                    <div className="card-top-container">
                      <div className="card-img-overlay">
                        <div className="card-header h1">Focused Artwork</div>
                        <div className="card-img-result">
                          {/* @ts-ignore */}
                          <ProgressiveImage
                            src={artwork.url}
                            placeholder={artwork.url_low_quality}
                          >
                            {(src) => (
                              <img
                                src={src}
                                alt="match_image"
                                role="img"
                                aria-label={`${artwork.title} by ${
                                  artwork.artist
                                }${
                                  artwork.culture
                                    ? `, ${artwork.culture}.`
                                    : "."
                                } ${artwork.visualDescription}`}
                              />
                            )}
                          </ProgressiveImage>
                        </div>
                        <div className="card-artist">{artwork.artist}</div>
                        <div className="card-title">{artwork.title}</div>
                      </div>
                    </div>
                    <div className="card-body" id="focussed-artwork-body">
                      <div className="share-wrapper">
                        {/* Language options button */}
                        <div className="language-dropdown-wrapper">
                          <div className="language-dropdown">
                            <LanguageDropdown
                              langOptions={langOptions}
                              selected={selectedLanguage}
                              onSelectLanguage={onSelectLanguage}
                            />
                          </div>
                        </div>

                        {/* Share options button for Barnes Collection objects */}
                        {/* TODO: determine if we want share button for special exhibition objects */}
                        {!result.data.specialExhibition && (
                          <Share
                            shareText={getTranslation("Result_page", "text_1")}
                            artwork={artwork}
                          />
                        )}
                      </div>

                      <div className="short-desc-container">
                        {/* Content blocks from Hygraph CMS */}
                        {artwork.content && (
                          <div className="card-content">
                            {artwork.content.map((c, index) => (
                              <ContentBlock
                                contentBlock={c.contentBlock}
                                key={index}
                              />
                            ))}
                          </div>
                        )}
                      </div>

                      {artwork.content &&
                        selectedLanguage.code !== LANGUAGE_EN && (
                          <div className="google-translate-disclaimer">
                            <span>Translated with </span>
                            <img src={google_logo} alt="google_logo" />
                          </div>
                        )}

                      <div className="card-info">
                        <table className="detail-table">
                          <tbody>
                            <tr>
                              <td className="text-left item-label">{`${getTranslation(
                                "Result_page",
                                "text_3"
                              )}:`}</td>
                              <td className="text-left item-info">
                                {artwork.artist}{" "}
                                {!artwork.unIdentified && artwork.nationality
                                  ? `(${artwork.nationality}, ${artwork.birthDate} - ${artwork.deathDate})`
                                  : ""}
                              </td>
                            </tr>
                            {artwork.unIdentified && (
                              <tr>
                                <td className="text-left item-label">
                                  {`${getTranslation(
                                    "Result_page",
                                    "text_10"
                                  )}:`}
                                </td>
                                <td className="text-left item-info">
                                  {artwork.culture}
                                </td>
                              </tr>
                            )}
                            <tr>
                              <td className="text-left item-label">{`${getTranslation(
                                "Result_page",
                                "text_4"
                              )}:`}</td>
                              <td className="text-left item-info">
                                {artwork.title}
                              </td>
                            </tr>
                            <tr>
                              <td className="text-left item-label">{`${getTranslation(
                                "Result_page",
                                "text_5"
                              )}:`}</td>
                              <td className="text-left item-info">
                                {artwork.displayDate}
                              </td>
                            </tr>
                            <tr>
                              <td className="text-left item-label">{`${getTranslation(
                                "Result_page",
                                "text_6"
                              )}:`}</td>
                              <td className="text-left item-info">
                                {artwork.medium}
                              </td>
                            </tr>
                            <tr>
                              <td className="text-left item-label">{`${getTranslation(
                                "Result_page",
                                "text_7"
                              )}:`}</td>
                              <td className="text-left item-info">
                                {artwork.dimensions}
                              </td>
                            </tr>
                            {!artwork.curatorialApproval && (
                              <tr>
                                <td className="text-left item-label">
                                  {`${getTranslation(
                                    "Result_page",
                                    "text_8"
                                  )}:`}
                                </td>
                                <td className="text-left item-info">
                                  {getTranslation("Result_page", "text_9")}
                                </td>
                              </tr>
                            )}
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div id="card-2__trigger" />
            </div>
            {showEmailForm ? (
              <div
                id="card-2"
                className={classnames(
                  "exhibition-obj__content__email",
                  "card",
                  {
                    open: formOpen,
                    completed: !formOpen && emailCaptured,
                  }
                )}
                onClick={handleClick}
              >
                <EmailForm
                  withStory={false}
                  isEmailScreen={false}
                  onSubmitEmail={handleSubmitEmail}
                  getTranslation={getTranslation}
                  getSize={(height) => null}
                  pointerEvents="auto"
                  handleClickScroll={(storyIndex, isStoryCard) => null}
                  alwaysFloatBtn={true}
                />
              </div>
            ) : (
              <ScanButton float={false} />
            )}
          </Fragment>
        )}
      </div>
    </div>
  );
};

export const ArtworkDefault = withTranslation<ArtworkDefaultProps>(
  ArtworkDefaultComponent
);
