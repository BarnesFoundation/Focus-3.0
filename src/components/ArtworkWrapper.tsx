import React, { Fragment, useEffect, useRef, useState } from "react";
import { useLocation } from "react-router-dom";
import classnames from "classnames";
import { isTablet } from "react-device-detect";
import { isAndroid } from "react-device-detect";

import { SearchRequestService } from "../services/SearchRequestService";
import * as constants from "../constants";
import withTranslation, {
  LanguageOptionType,
  WithTranslationState,
} from "./withTranslation";
import {
  constructResultAndInRoomSlider,
  constructStory,
} from "../helpers/artWorkHelper";
import {
  ArtworkObject,
  ArtWorkRecordsResult,
  StoryResponse,
} from "../types/payloadTypes";
import { useLocalStorage } from "../hooks/useLocalStorage";
import { EmailForm } from "./EmailForm";
import ProgressiveImage from "react-progressive-image";
import { LanguageDropdown } from "./LanguageDropdown";
import { Share } from "./Share";
import { ContentBlock } from "./ContentBlock";
import google_logo from "../images/google_translate.svg";
import { ScanButton } from "./ScanButton";
import Artwork from "./Artwork";
import { ExhibitionObject } from "./ExhibitionObject";

export const ArtworkWrapperComponent: React.FC<WithTranslationState> = ({
  getSelectedLanguage,
}) => {
  // Initialize the search request services
  const sr = new SearchRequestService();
  // Component state
  const [imageId, setImageId] = useState<string>();
  const [artwork, setArtwork] = useState<ArtworkObject["artwork"]>();
  const [result, setResult] = useState<ArtWorkRecordsResult>();
  const [stories, setStories] = useState<StoryResponse>();
  const [imgLoaded, setImgLoaded] = useState(false);
  const [emailCaptured, setEmailCaptured] = useState(false);
  const [showEmailForm, setShowEmailForm] = useState(true);
  const [emailCaptureAck, setEmailCaptureAck] = useState(false);
  const [selectedLanguage, setSelectedLanguage] = useState<LanguageOptionType>({
    name: "English",
    code: "En",
    selected: true,
  });

  const location = useLocation();
  const { getLocalStorage } = useLocalStorage();

  const [formOpen, setFormOpen] = useState(false);
  const startYCard2 = useRef<number>();
  const startYTrigger = useRef<number>();

  const setupStory = async (imageId) => {
    const storyInformation = await sr.getStoryItems(imageId);
    return constructStory(storyInformation);
  };

  // Set state and get object data on component initialization
  useEffect(() => {
    const componentWillMount = async () => {
      let artworkInfo: ArtWorkRecordsResult;
      let storyResponse: StoryResponse;
      let imageId: string;

      // Check if result data is stored in location state from camera component
      if (location.state?.result) {
        imageId = location.state.result.data.records[0].id;
        artworkInfo = location.state.result;
      } else {
        // If result data is not in state, get image ID from path and get data from server
        const [_n, path, id] = location.pathname.split("/");
        imageId = id;

        if (path === "artwork") {
          [artworkInfo, storyResponse] = await Promise.all([
            sr.getArtworkInformation(imageId),
            setupStory(imageId),
          ]);
        } else if (path === "exhibition") {
          artworkInfo = await sr.getSpecialExhibitionObject(imageId);
        }
      }

      const { artwork, roomRecords } = constructResultAndInRoomSlider(
        artworkInfo,
        isTablet
      );

      setResult(artworkInfo);
      setArtwork(artwork);
      setStories(storyResponse);
      setImageId(imageId);

      const emailRecorded = getLocalStorage(constants.SNAP_USER_EMAIL) !== null;
      setEmailCaptured(emailRecorded);
      setShowEmailForm(!emailRecorded);
      setEmailCaptureAck(emailRecorded);
      setSelectedLanguage((await getSelectedLanguage())[0]);
    };

    componentWillMount();
  }, []);

  return (
    <Fragment>
      {result &&
        artwork &&
        (result.data.showStory ? (
          <Artwork />
        ) : (
          <ExhibitionObject
            initArtwork={artwork}
            initResult={result}
            imageId={imageId}
          />
        ))}
    </Fragment>
  );
};

export const ArtworkWrapper = withTranslation(ArtworkWrapperComponent);
