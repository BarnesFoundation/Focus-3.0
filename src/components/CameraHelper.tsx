export const loadImage = (url) => {
  return new Promise((resolve, reject) => {
    // Create a new image
    const image = new Image();

    // Bind an event listener on the load to call the `resolve` function
    image.onload = resolve;

    // If the image fails to be downloaded, we don't want the whole system to collapse so we `resolve` instead of `reject`, even on error
    image.onerror = resolve;
    image.src = url;
  });
};

export const resetScanCounter = () => {};
