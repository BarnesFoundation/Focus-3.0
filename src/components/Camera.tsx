import React, { Component } from "react";
import { compose } from "redux";
import { cropPhoto } from "./CameraHelper";
import * as constants from "../constants";
import withOrientation from "./withOrientation";
import withTranslation from "./withTranslation";
import { shouldLogPermissionGrantTime } from "../helpers/googleAnalyticsHelpers";
import { NoMatchOverlay } from "./NoMatchOverlay";

const DISABLE_ZOOM = "DISABLE_ZOOM";
const ENABLE_ZOOM = "ENABLE_ZOOM";

type CameraProps = {
  sessionYieldedMatch: boolean;
  beginScanning: () => void;
  snapAttempts: number;
  shouldBeScanning: boolean;
  processImageCapture: (imageBlob: any) => Promise<void>;
};

type CameraState = {
  videoStream: MediaStream;
  frontCamera: boolean;
  snapAttempts: CameraProps["snapAttempts"];
  error?: string;
};

class Camera extends Component<CameraProps, CameraState> {
  cropRectangle: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
  scan: NodeJS.Timer;
  captureCounter: number;
  previewCanvas: HTMLCanvasElement;
  video: HTMLVideoElement;
  vpreview: HTMLCanvasElement;
  initVideo: NodeJS.Timeout;
  drawPreview: NodeJS.Timeout;
  tempCanvas: HTMLCanvasElement;
  tempCtx: CanvasRenderingContext2D;

  constructor(props) {
    super(props);

    // Set state variables
    this.state = {
      videoStream: null,
      frontCamera: false,
      snapAttempts: this.props.snapAttempts,
    };
  }

  resetSnapCounter = () => {
    let lastSnapTimestamp = parseInt(
      localStorage.getItem(constants.SNAP_LAST_TIMESTAMP)
    );

    if (lastSnapTimestamp) {
      let ttl =
        lastSnapTimestamp + constants.SNAP_COUNT_RESET_INTERVAL - Date.now();

      if (ttl <= 0 && this.state.snapAttempts > 0) {
        localStorage.removeItem(constants.SNAP_ATTEMPTS);
        this.setState({ snapAttempts: 0 });
      }
    }
  };

  /** Captures a single scan and returns blob of the scan */
  captureSingleScan = async () => {
    this.captureCounter += 1;

    if (this.captureCounter < 10) {
      // Get image in canvas
      let canvas = this.previewCanvas;

      if (!canvas) return null;

      // Get the blob from canvas
      const imageBlob = await new Promise((resolve) => {
        canvas.toBlob(async (blob) => resolve(blob), "image/jpeg", 0.75);
      });

      this.props.processImageCapture(imageBlob);
    }
  };

  playVideo = async () => {
    if (this.video) {
      try {
        await this.video.play();

        const track = this.state.videoStream.getVideoTracks()[0];

        // Begin scanning
        requestAnimationFrame(this.drawVideoPreview);
      } catch (error) {
        console.log("Cannot play video stream", error);
      }
    }
  };

  async componentDidMount() {
    // Disable zoom
    this.pinchZoomModifier(DISABLE_ZOOM);

    const previewBox = this.vpreview.getBoundingClientRect();
    const { x, y, width, height } = previewBox;
    this.cropRectangle = {
      x: Math.floor(x),
      y: Math.floor(y),
      width: Math.floor(width),
      height: Math.floor(height),
    };

    // Fetch the device camera
    try {
      const startTime = Date.now();
      const videoStream = await navigator.mediaDevices.getUserMedia({
        video: {
          facingMode: this.state.frontCamera ? "user" : "environment",
          width: 1920,
          height: 1080,
        },
      });

      // Log the permission grant time if it took more than 900 ms
      // TODO - fix this
      // shouldLogPermissionGrantTime(startTime);

      this.setState({ videoStream }, () => {
        // When video is able to be captured
        if (this.state.videoStream) {
          this.setupForCapturing();
          this.resetSnapCounter();
        }
      });
    } catch (error) {
      this.setState({ error: "An error occurred accessing the device camera" });
    }
  }

  /* Prevents zoom */
  disablePinchZoom = (event) => {
    event.preventDefault();
  };

  /* Beginning with iOS 10, the "user-scalable=no" attribute is no longer supported. This function will disable page zoom on-pinch for the camera page */
  pinchZoomModifier = (action) => {
    const cameraContainer = document.querySelector(".camera");

    if (action === DISABLE_ZOOM) {
      cameraContainer.addEventListener(
        "touchmove",
        this.disablePinchZoom,
        false
      );
    }

    if (action === ENABLE_ZOOM) {
      cameraContainer.removeEventListener("touchmove", this.disablePinchZoom);
    }
  };

  stopVideo = () => {
    if (this.initVideo) {
      clearTimeout(this.initVideo);
    }
  };

  stopPreview = () => {
    if (this.drawPreview) {
      clearTimeout(this.drawPreview);
    }
  };

  setupForCapturing = async () => {
    this.video.srcObject = this.state.videoStream;
    this.stopVideo();
    this.initVideo = setTimeout(() => {
      this.playVideo();
    }, 50);
    this.captureCounter = 0;

    this.scan = await new Promise((resolve) => {
      setTimeout(() => {
        const scan = setInterval(this.captureSingleScan, 1000 / 3);
        resolve(scan);
      }, 500);
    });
  };

  async componentDidUpdate(previousProps) {
    console.log(
      `previous shouldBeScanning: ${previousProps.shouldBeScanning} current shouldBeScanning: ${this.props.shouldBeScanning}, current sessionYieldedMatch: ${this.props.sessionYieldedMatch}`
    );

    if (
      !previousProps.shouldBeScanning &&
      this.props.shouldBeScanning == true
    ) {
      await this.setupForCapturing();
    }

    if (this.props.shouldBeScanning == false) {
      clearInterval(this.scan);
    }
  }

  componentWillUnmount() {
    // Stop drawing the video
    this.stopVideo();
    this.stopPreview();

    // Turn off video capture
    this.video.pause();
    this.video.removeAttribute("src");
    this.video.load();

    // Re-enable zoom
    this.pinchZoomModifier(ENABLE_ZOOM);
  }

  /** Return a temporary canvas and context for drawing the video */
  getTempPreviewCanvas = () => {
    const rect = this.video.getBoundingClientRect();
    if (!this.tempCanvas) {
      let tempCanvas = document.createElement("canvas");
      let tempCtx = tempCanvas.getContext("2d", { alpha: false });
      tempCanvas.width = rect.width;
      tempCanvas.height = rect.height;

      this.tempCanvas = tempCanvas;
      this.tempCtx = tempCtx;
    }
    const { tempCanvas, tempCtx } = this;
    return { tempCanvas, tempCtx };
  };

  /** Draws the portion of the video visible within the preview onto a canvas, in a loop. */
  drawVideoPreview = () => {
    try {
      if (!this.vpreview /*  || this.props.sessionYieldedMatch === null */) {
        return null;
      }

      let { tempCanvas, tempCtx } = this.getTempPreviewCanvas();

      let previewCanvas = this.vpreview;
      let previewContext = previewCanvas.getContext("2d", { alpha: false });
      previewCanvas.width = this.cropRectangle.width;
      previewCanvas.height = this.cropRectangle.height;

      tempCtx.clearRect(0, 0, tempCanvas.width, tempCanvas.height);
      tempCtx.drawImage(this.video, 0, 0, tempCanvas.width, tempCanvas.height);

      previewContext.clearRect(0, 0, previewCanvas.width, previewCanvas.height);
      previewContext.drawImage(
        tempCanvas,
        this.cropRectangle.x,
        this.cropRectangle.y,
        this.cropRectangle.width,
        this.cropRectangle.height,
        0,
        0,
        previewCanvas.width,
        previewCanvas.height
      );

      this.previewCanvas = previewContext.canvas;

      this.stopPreview();
      this.drawPreview = setTimeout(() => {
        requestAnimationFrame(this.drawVideoPreview);
      }, 10);
    } catch (error) {
      console.log("Error setting camera preview");
      // TODO: re-enable the GA tracking
      // ga("send", {
      //   hitType: "event",
      //   eventCategory: constants.GA_EVENT_CATEGORY.CAMERA,
      //   eventAction: constants.GA_EVENT_ACTION.SCAN,
      //   eventLabel: constants.GA_EVENT_LABEL.SCANNER_MOUNT_FAILURE,
      // });
    }
  };

  render() {
    const { beginScanning, sessionYieldedMatch, shouldBeScanning } = this.props;

    let videoStyle = { filter: `blur(25px)`, transform: `scale(1.2)` };

    return (
      <div className="camera">
        {/* If scanning should be occurring */}
        {
          <div>
            {/* Get video */}
            {
              <video
                id="video"
                ref={(c) => (this.video = c)}
                width="100%"
                autoPlay
                playsInline
                muted
                style={videoStyle}
              />
            }

            {/* Show the video preview if an unsuccessful attempt has not occurred */}
            {sessionYieldedMatch !== false && (
              <canvas id="video-preview" ref={(el) => (this.vpreview = el)} />
            )}

            {/* If there was an unsuccessful attempt, transition into the no result found */}
            <NoMatchOverlay
              displayOverlay={
                shouldBeScanning === false && sessionYieldedMatch === false
              }
              handleScan={() => beginScanning()}
            />
          </div>
        }
      </div>
    );
  }
}

export default compose(withOrientation, withTranslation)(Camera);
