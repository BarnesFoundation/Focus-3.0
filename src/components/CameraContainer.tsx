import React, { Component } from "react";
import { withRouter, History } from "react-router-dom";
import { compose } from "redux";

import posed from "react-pose";

import Camera from "./Camera";
import { StorableSearch, ImageSearchResponse } from "../classes/searchResponse";
import { SearchRequestService } from "../services/SearchRequestService";
import { loadImage } from "./CameraHelper";
import * as constants from "../constants";
import { ARTWORK, EXHIBITION } from "../constants/routes";
import withTranslation, { WithTranslationState } from "./withTranslation";

const Container = posed.div({
  enter: { opacity: 1 },
  exit: { opacity: 0 },
});

type CameraContainerProps = {
  history: History;
} & WithTranslationState;

type CameraContainerState = {
  scanSeqId: any;
  shouldBeScanning: boolean;
  sessionYieldedMatch: boolean;
};

export class CameraContainerComponent extends Component<
  CameraContainerProps,
  CameraContainerState
> {
  sr: SearchRequestService;
  snapAttempts: number;
  responseCounter: number;

  constructor(props) {
    super(props);

    this.sr = new SearchRequestService();
    this.snapAttempts =
      parseInt(localStorage.getItem(constants.SNAP_ATTEMPTS)) || 0;

    this.state = {
      scanSeqId: null,
      shouldBeScanning: null,
      sessionYieldedMatch: null,
    };
  }

  /** Starts a scanning session over a 3-second duration */
  beginScanning = () => {
    // Update the snap attempts with this scan as a single attempt
    localStorage.setItem(
      constants.SNAP_ATTEMPTS,
      (this.snapAttempts + 1).toString()
    );
    localStorage.setItem(constants.SNAP_LAST_TIMESTAMP, Date.now().toString());

    // Reset scans taken and stop any existing scanning
    this.responseCounter = 0;
    this.snapAttempts = this.snapAttempts + 1;
    this.setState({
      shouldBeScanning: true,
      scanSeqId: Date.now(),
      sessionYieldedMatch: null,
    });
  };

  /** Executes steps for completing a scan from an image capture to determining a match */
  processImageCapture = async (imageBlob) => {
    // Only process image capture when we haven't determined if session yielded a match
    if (this.state.sessionYieldedMatch === null && this.responseCounter <= 9) {
      // Prepare and send the request to Catchoom for a response
      const imageSearchRequestConfig = this.sr.prepareRequest(
        imageBlob,
        this.state.scanSeqId
      );
      const imageSearchResponse = await this.sr.submitImageSearchRequest(
        imageSearchRequestConfig
      );

      const { data } = imageSearchRequestConfig;
      const { searchWasSuccessful, searchTime } = imageSearchResponse;
      let searchResultToStore,
        elasticSearchResponse = null;

      // Increment the response counter
      this.responseCounter++;

      // If search was successful
      if (searchWasSuccessful && this.state.shouldBeScanning) {
        this.setState({ shouldBeScanning: false, sessionYieldedMatch: true });

        // Get the identified image information
        const identifiedItem = imageSearchResponse.responsePayload.results[0];

        const identifiedImageInformation = await this.sr.processIdentifiedItem(
          identifiedItem
        );

        const { esResponse, referenceImageUrl } = identifiedImageInformation;

        searchResultToStore = new StorableSearch(
          data,
          searchWasSuccessful,
          referenceImageUrl,
          esResponse,
          searchTime
        );
        elasticSearchResponse = esResponse;

        if (elasticSearchResponse) {
          this.completeImageSearchRequest(elasticSearchResponse);
        } else {
           // The ES response for special exhibition objects can be undefined if no content has bee added to Hygraph for the object,
          //  In that case we will want to end scanning by updating the state to show that the search did not yield a match
          this.setState({
            sessionYieldedMatch: false,
          });
        }
      }

      // Otherwise, when it's not successful
      else {
        searchResultToStore = new StorableSearch(
          data,
          searchWasSuccessful,
          null,
          elasticSearchResponse,
          searchTime
        );
      }

      // Store the result, regardless of success or not
      this.sr.storeSearchedResult(searchResultToStore);

      // If we've received 9 responses and haven't completed the search, the session yielded no match
      if (
        this.responseCounter === 9 &&
        this.state.sessionYieldedMatch == null &&
        this.state.shouldBeScanning
      ) {
        // Updates the state to show that the search did not yield a match and scanning should end
        this.setState({ shouldBeScanning: false, sessionYieldedMatch: false });
      }
    }
  };

  /** Processes the completion of successful image search */
  completeImageSearchRequest(response) {
    const { width } = screen;

    // Get the record and art url from it
    const record = response["data"]["records"][0];
    const { art_url: artUrl, id } = record;
    const specialExhibition: boolean = response["data"]["specialExhibition"];

    // Load the image background first so that it gets cached for faster displaying
    const matchImage = loadImage(`${artUrl}?w=${width - 80}`);
    const matchImageBg = loadImage(
      `${artUrl}?q=0&auto=compress&crop=faces,entropy&fit=crop&w=${width}`
    );

    Promise.all([matchImage, matchImageBg]).then(() => {
      // Navigate to the artwork or exhibition page
      this.props.history.push({
        pathname: `${specialExhibition ? EXHIBITION : ARTWORK}/${id}`,
        state: { result: response },
      });
    });
  }

  componentDidMount() {
    this.beginScanning();
  }

  render() {
    const { processImageCapture, snapAttempts, beginScanning } = this;
    const { sessionYieldedMatch, shouldBeScanning } = this.state;

    return (
      <Container className="camera-container" initialPose="exit" pose="enter">
        <Camera
          // @ts-ignore
          processImageCapture={processImageCapture}
          sessionYieldedMatch={sessionYieldedMatch}
          beginScanning={beginScanning}
          snapAttempts={snapAttempts}
          shouldBeScanning={shouldBeScanning}
        />
      </Container>
    );
  }
}

export default compose(withRouter, withTranslation)(CameraContainerComponent);
