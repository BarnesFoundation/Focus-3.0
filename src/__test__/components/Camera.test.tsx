import React from "react";
import { render, cleanup } from "@testing-library/react";

import { CameraComponent } from "../../components/Camera";

afterEach(cleanup);

describe("Camera", () => {
  it("should match the snapshot when session yielded match is true", () => {
    const { container } = render(
      <CameraComponent
        sessionYieldedMatch={true}
        beginScanning={() => null}
        snapAttempts={0}
        shouldBeScanning={false}
        processImageCapture={async (blog) => null}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot when session yielded match is false", () => {
    const { container } = render(
      <CameraComponent
        sessionYieldedMatch={false}
        beginScanning={() => null}
        snapAttempts={0}
        shouldBeScanning={false}
        processImageCapture={async (blog) => null}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot when session should be scanning is true", () => {
    const { container } = render(
      <CameraComponent
        sessionYieldedMatch={false}
        beginScanning={() => null}
        snapAttempts={0}
        shouldBeScanning={true}
        processImageCapture={async (blog) => null}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
