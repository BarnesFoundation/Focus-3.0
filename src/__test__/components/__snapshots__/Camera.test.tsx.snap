// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera should match the snapshot when session should be scanning is true 1`] = `
<div>
  <div
    class="camera"
  >
    <div>
      <video
        autoplay=""
        id="video"
        playsinline=""
        style="filter: blur(25px); transform: scale(1.2);"
        width="100%"
      />
      <div />
    </div>
  </div>
</div>
`;

exports[`Camera should match the snapshot when session yielded match is false 1`] = `
<div>
  <div
    class="camera"
  >
    <div>
      <video
        autoplay=""
        id="video"
        playsinline=""
        style="filter: blur(25px); transform: scale(1.2);"
        width="100%"
      />
      <div />
    </div>
  </div>
</div>
`;

exports[`Camera should match the snapshot when session yielded match is true 1`] = `
<div>
  <div
    class="camera"
  >
    <div>
      <video
        autoplay=""
        id="video"
        playsinline=""
        style="filter: blur(25px); transform: scale(1.2);"
        width="100%"
      />
      <canvas
        id="video-preview"
      />
      <div />
    </div>
  </div>
</div>
`;
